doc/authentication.md

```diff
@@ -0,0 +1,47 @@
+
+Cockpit Authentication
+================================
+
+Cockpit authorizes users by looking at the Authorization header. The contents
+of that header are sent to an external command that is then responsible to:
+
+ * verify the given credentials
+ * setup an appropriate session and enviroment based on those credentials
+ * launch a bridge that can communicate with cockpit-ws on stdin and stdout.
+
+Cockpit ships with the cockpit-session command that is able to handle basic and gssapi
+authentication. Cockpit supports per auth scheme configuration so that alternate commands
+can be used.
+
+To configure an auth scheme add a section to cockpit.conf for that scheme. For example
+
+```
+[bearer]
+command = example-verify-token
+base64-decode = false
+```
+
+cockpit-ws will then call the given command whenever it receives a Bearer token in a
+Authentication header.
+
+The command will be called with two arguments
+
+ * type: the auth scheme that was included in the the Authorization http header
+ * remote host: The remote host to use with pam.
+
+cockpit-ws will then send the base64 decoded contents of the Authorization http header,
+without the auth scheme, on fd #3 to the command. You can skip the base64 decoding by
+setting ```base64-decode = false``` in your cockpit.conf file.
+
+Once the command has processed the credentials it MUST write a JSON response to
+fd #3. The response should contain the following fields
+
+ * response-code: This field is always required.
+  * -1 indicates in internal error.
+  * 0 indicates success.
+  * Any other number is treated as an error
+ * error: Should be present on object where the response-code is not 0. Values of ```authentication-failed```, ```authentication-unavailable``` or ```permission-denied``` are translated to the appropriate cockpit error codes. Any other values are treated as generic errors
+ * error-details: Optional text with more details about the error.
+ * user: Should be present on any object where the response-code is 0
+
+Once the response has been sent fd #3 should be closed and a cockpit-bridge launched.
```

Comment:
Lets call this a "bridge" since it's not always ```cockpit-bridge``` ... and say: a bridge should be launched speaking the cockpit protocol on stdin and stdout.